crates/bevy_ecs/src/query/mod.rs

```diff
@@ -137,6 +137,182 @@ mod tests {
+        assert_eq!(
+            a_query_with_b.iter_combinations::<3>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(a_query_with_b.iter_combinations::<4>(&world).count(), 0);
+        assert_eq!(
+            a_query_with_b.iter_combinations::<4>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(a_query_with_b.iter_combinations::<5>(&world).count(), 0);
+        assert_eq!(
+            a_query_with_b.iter_combinations::<5>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(a_query_with_b.iter_combinations::<1024>(&world).count(), 0);
+        assert_eq!(
+            a_query_with_b.iter_combinations::<1024>(&world).size_hint(),
+            (0, Some(0))
+        );
+
+        let mut a_query_without_b = world.query_filtered::<&A, Without<B>>();
+        assert_eq!(a_query_without_b.iter_combinations::<0>(&world).count(), 0);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<0>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(a_query_without_b.iter_combinations::<1>(&world).count(), 3);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<1>(&world).size_hint(),
+            (0, Some(3))
+        );
+        assert_eq!(a_query_without_b.iter_combinations::<2>(&world).count(), 3);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<2>(&world).size_hint(),
+            (0, Some(3))
+        );
+        assert_eq!(a_query_without_b.iter_combinations::<3>(&world).count(), 1);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<3>(&world).size_hint(),
+            (0, Some(1))
+        );
+        assert_eq!(a_query_without_b.iter_combinations::<4>(&world).count(), 0);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<4>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(a_query_without_b.iter_combinations::<5>(&world).count(), 0);
+        assert_eq!(
+            a_query_without_b.iter_combinations::<5>(&world).size_hint(),
+            (0, Some(0))
+        );
+        assert_eq!(
+            a_query_without_b.iter_combinations::<1024>(&world).count(),
+            0
+        );
+        assert_eq!(
+            a_query_without_b
+                .iter_combinations::<1024>(&world)
+                .size_hint(),
+            (0, Some(0))
+        );
+
+        let values: Vec<[&A; 2]> = a_query_without_b.iter_combinations(&world).collect();
+        assert_eq!(
+            values,
+            vec![[&A(2), &A(3)], [&A(2), &A(4)], [&A(3), &A(4)],]
+        );
+
+        let values: Vec<[&A; 3]> = a_query_without_b.iter_combinations(&world).collect();
+        assert_eq!(values, vec![[&A(2), &A(3), &A(4)],]);
+
+        let mut query = world.query_filtered::<&A, Or<(With<A>, With<B>)>>();
+        let values: Vec<[&A; 2]> = query.iter_combinations(&world).collect();
```

Comment:
Do we make any guarantees about `iter_combinations` order?
I think we should have a `.sort_unstable` here ideally; I don't want our tests to rely on any specific ordering.
OTOH I struggle to think of a scenario where we might want to 'break' the order in future

Reply:
I agree that we shouldn't rely on specific ordering at all. I originally copied some parts of the test from the other test (`query_iter_combinations`) and I was confused why Vec is used instead of HashSet.

I'll change some parts of the test I've written to use HashSet instead, but I will not change the other test to keep the scope of this PR narrow.